jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/impl/NodeInstanceFactoryRegistry.java

```diff
@@ -67,89 +67,149 @@
-        	NodeInstanceFactory result = this.registry.get( clazz );
-        	if (result != null) {
-        		return result;
-        	}
-        	clazz = clazz.getSuperclass();
+            NodeInstanceFactory result = this.get(clazz);
+            if (result != null) {
+                return result;
+            }
+            clazz = clazz.getSuperclass();
         }
         return null;
     }
+
+    private NodeInstanceFactory get(Class<?> clazz) {
+        // hard wired nodes:
+        if (RuleSetNode.class == clazz) {
+            return factory(RuleSetNodeInstance::new);
+        }
+        if (Split.class == clazz) {
+            return factory(SplitInstance::new);
+        }
+        if (Join.class == clazz) {
+            return factoryOnce(JoinInstance::new);
+        }
+        if (StartNode.class == clazz) {
+            return factory(StartNodeInstance::new);
+        }
+        if (EndNode.class == clazz) {
+            return factory(EndNodeInstance::new);
+        }
+        if (MilestoneNode.class == clazz) {
+            return factory(MilestoneNodeInstance::new);
+        }
+        if (SubProcessNode.class == clazz) {
+            return factory(SubProcessNodeInstance::new);
+        }
+        if (ActionNode.class == clazz) {
+            return factory(ActionNodeInstance::new);
+        }
+        if (WorkItemNode.class == clazz) {
+            return factory(WorkItemNodeInstance::new);
+        }
+        if (TimerNode.class == clazz) {
+            return factory(TimerNodeInstance::new);
+        }
+        if (FaultNode.class == clazz) {
+            return factory(FaultNodeInstance::new);
+        }
+        if (EventSubProcessNode.class == clazz) {
+            return factory(EventSubProcessNodeInstance::new);
+        }
+        if (CompositeNode.class == clazz) {
+            return factory(CompositeNodeInstance::new);
+        }
+        if (CompositeContextNode.class == clazz) {
+            return factory(CompositeContextNodeInstance::new);
+        }
+        if (HumanTaskNode.class == clazz) {
+            return factory(HumanTaskNodeInstance::new);
+        }
+        if (ForEachNode.class == clazz) {
+            return factory(ForEachNodeInstance::new);
+        }
+        if (EventNode.class == clazz) {
+            return factory(EventNodeInstance::new);
+        }
+        if (StateNode.class == clazz) {
+            return factory(StateNodeInstance::new);
+        }
+        if (DynamicNode.class == clazz) {
+            return factory(DynamicNodeInstance::new);
+        }
+        if (BoundaryEventNode.class == clazz) {
+            return factory(BoundaryEventNodeInstance::new);
+        }
+        if (CatchLinkNode.class == clazz) {
+            return factory(
+                    CatchLinkNodeInstance::new);
+        }
+        if (ThrowLinkNode.class == clazz) {
+            return factory(
+                    ThrowLinkNodeInstance::new);
+        }
+        return this.registry.get(clazz);
```

Comment:
we retain extensibility but at the same time avoid creating hashmaps if necessary. It is true this is O(n) while HashMaps are O(1), but for small Ns odds are in favor of linear scan.

Notice this can be run entirely in SVM pre-compile phase (we'll discuss this separately)